src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -553,35 +553,46 @@ class Wavefunction
     /// \pre: Each qubit, listed in `q`, must be unentangled and in state |0>.
     /// Place qubits, listed in `q` into superposition of basis vectors with provided `amplitudes`, where the order of
     /// qubits in array `q` defines the order of the basis vectors (little endian).
-    /// NB: at the moment supported only total state injection on all currently allocated qubits
     void inject_state(const std::vector<logical_qubit_id>& qubits, const std::vector<ComplexType>& amplitudes)
     {
         assert((static_cast<size_t>(1) << qubits.size()) == amplitudes.size());
 
-        if (num_qubits_ != qubits.size()) throw std::runtime_error("Subsystem state injection not supported (yet)");
-
         flush();
 
         // Check prerequisites.
         for (logical_qubit_id q : qubits)
         {
-            if (!kernels::isclassical(wfn_, get_qubit_position(q))
-                || kernels::getvalue(wfn_, get_qubit_position(q)) != 0)
+            if (!kernels::isclassical(wfn_, get_qubit_position(q)) ||
+                kernels::getvalue(wfn_, get_qubit_position(q)) != 0)
             {
                 throw std::runtime_error("Cannot prepare state of entangled qubits or if they are not in state |0>");
             }
         }
 
-        // Reorder the qubits to the positions that match the wave function provided by the user.
-        for (unsigned i = 0; i < qubits.size(); i++)
+        if (qubits.size() == num_qubits_)
         {
-            qubitmap_[qubits[i]] = i;
+            // For full state injection we can simply copy the user's wave function into our store and reorder the
+            // positions map, but is it actually/meaningfully faster to special case total state injection?
+            for (unsigned i = 0; i < qubits.size(); i++)
+            {
+                qubitmap_[qubits[i]] = i;
+            }
+            wfn_.assign(amplitudes.begin(), amplitudes.end());
         }
-
-        // For full state injection we can simply copy the user's wave function into our store.
-        for (size_t i = 0; i < wfn_.size(); i++)
+        else
         {
-            wfn_[i] = amplitudes[i];
+            const size_t num_states = wfn_.size();
+            std::vector<positional_qubit_id> positions = get_qubit_positions(qubits);
+            const size_t mask = kernels::make_mask(positions);
+            WavefunctionStorage wfn_new(num_states);
+
+            for (size_t basis_index = 0; basis_index < num_states; basis_index++)
```

Comment:
If I understand correctly, the approach here is to:

- write out the state as Σᵢ (𝑎ᵢ |𝑖⟩ ⊗ |0⟩), where the first register is the qubits not being injected and where the second register should have the state being injected. Each 𝑎ᵢ can be found using `wfn_[detail::set_register(...)]`.
- inject the state Σᵢ 𝑎ᵢ |𝑖⟩ ⊗ |φ⟩ = Σᵢ,ⱼ (𝑎ᵢ 𝑏ⱼ |𝑖⟩ ⊗ |𝑗⟩), where |φ⟩ = Σⱼ 𝑏ⱼ |𝑗⟩ is the state to be injected and where {𝑏ⱼ}ⱼ is given by `amplitudes`. Since each term |𝑖⟩ ⊗ |𝑗⟩ is a computational basis state, that can be injected by setting the amplitudes of `wfn_new` directly for each index (𝑖, 𝑗) to be 𝑎ᵢ 𝑏ⱼ.

Reply:
Hmm, the bechmarking results are fairly surprising if I understand them correctly. They are not consistent with my previous state preparation experience. Can we do an experiment based on an equivalent of the following pseudo-code:
double[] v = new double[2^20]; double s2 = 0.0; for (size_t ix = 0; ix < 2^20; ix++) { 
double x= (double) rand(); s2+=x*x; v[ix]=x;} double nu = 1.0/sqrt(s2); for (size_t jx = 0; jx < 2^20; jx++) v[ jx]*=nu;}
After which we (a) feed v into the StatePreparation API and (b) feed v into the state injection API 
?
I would like to to understand where and how much time is spent in each scenario.

Reply:
I believe I figured out why the micro benchmarks on the native side were skewed: the simulator was caching the quantum operations so the wave function wasn't actually being updated. After forcing the flush the numbers are much closer to the expectations:

For creating the equal superposition of all basis vectors, the injection performs similar to the quantum preparation on systems with small number of qubits (< 10) and about x5 faster for systems with 20 qubits. Partial state injection yields smaller perf boost and might be worse than quantum state preparation when injecting a state on small subsystem into a large system. Basically, we might need heuristics at the Q# library level when deciding whether to inject or to prepare.

I haven't done the plumbing yet from Q# to native to measure it but the native micro benchmarks look good to me now and I think we should proceed with this change.